src/lib.rs

```diff
@@ -149,23 +149,92 @@ impl Key for &str {
+/// (and be supported by the intended [`serde::Serializer`]/[`serde::Deserializer`] backend)
+/// or implement [Miniconf].
+///
+/// For each field, the Miniconf recursion depth is configured through the `#[miniconf(defer(Y))]` attribute,
+/// with `Y = 1` being the implied default when using `#[miniconf(defer)]` and `Y = 0` invalid.
+/// If the attribute is not present, the field is a leaf and accessed only through its
+/// [`serde::Serialize`]/[`serde::Deserialize`] implementation.
+/// With the attribute present the field is accessed through its [`Miniconf<Y>`] implementation with the given
+/// recursion depth.
+///
+/// Homogeneous [core::array]s can be made accessible either
+/// 1. as a single leaf in the tree like other serde-capable items, or
+/// 2. by item through their numeric indices (with the attribute `#[miniconf(defer(1))]`), or
+/// 3. exposing a sub-tree per item with `#[miniconf(defer(D))]` and `D >= 2`.
+///
+/// `Option` is used
+/// 1. as a leaf like a standard `serde` Option, or
+/// 2. with `#[miniconf(defer(1))]` to support a leaf value that may be absent (masked) at runtime.
+/// 3. with `#[miniconf(defer(D))]` and `D >= 2` to support masking sub-trees at runtime.
+///
+/// ## Bounds on generics
+///
+/// The macro adds bounds to generic types of the struct it is acting on.
+/// E.g. If a generic type parameter `T` of the struct `S<T>`is used as a type parameter to a
+/// field type `a: F1<F2<T>>` the type `T` will be considered to reside at depth `X = 2` (as it is
+/// within `F2` which is within `F1`) and the following bounds will applied:
+///
+/// * With the `#[miniconf]` attribute not present, `T` will receive bounds `Serialize + DeserializeOwned`.
+/// * With `#[miniconf(defer(Y))]`, and `Y - X < 1` it will also receive bounds `Serialize + DeserializeOwned`.
+/// * For `Y - X >= 1` it will receive the bound `T: Miniconf<Y - X>`.
+///
+/// This behavior is upheld by and compatible with all implementations in this crate. It is only violated
 /// when deriving `Miniconf` for a struct that (a) forwards its own type parameters as type
 /// parameters to its field types, (b) uses `Miniconf` on those fields, and (c) those field
-/// types use their type parameters at other levels than `Miniconf<Y - 1>`.
+/// types use their type parameters at other levels than `Miniconf<Y - 1>`. See the
+/// `test_derive_macro_bound_failure` test in `tests/generics.rs`.
 ///
-/// The recursion depth `Y` is an upper bound of the maximum key length
-/// (the depth/height of the tree).
+/// # Example
+///
+/// ```rust
+/// use miniconf::Miniconf;
+///
+/// #[derive(Miniconf)]
+/// struct Nested {
+///     #[miniconf(defer)]
+///     data: [u32; 2],
+/// }
+/// #[derive(Miniconf)]
+/// struct Settings {
+///     // Accessed with path `nested/data/0` or `nested/data/1`
+///     #[miniconf(defer(2))]
+///     nested: Nested,
+///
+///     // Accessed with path `external`
```

Comment:
```suggestion
///     // Accessed with path `/nested/data/0` or `/nested/data/1`
///     #[miniconf(defer(2))]
///     nested: Nested,
///
///     // Accessed with path `/external`
```

Don't we enforce leading prefixes now?

Reply:
Everything before the first / is ignored. This only applies in json_core
Added doc about that there.